--- a/os_dep/linux/ioctl_cfg80211.c
+++ b/os_dep/linux/ioctl_cfg80211.c
@@ -3984,7 +3984,7 @@ static int rtw_cfg80211_add_monitor_if(_
 	mon_ndev->name[IFNAMSIZ - 1] = 0;
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 9))
 	mon_ndev->priv_destructor = rtw_ndev_destructor;
-	mon_ndev->needs_free_netdev = true;
+	mon_ndev->needs_free_netdev = false;
 #else
 	mon_ndev->destructor = rtw_ndev_destructor;
 #endif
